### ISSUE TITLE
psycopg2-static not working on AWS Lambda

### ISSUE THREAD

**Issue by dsvensson** [psycopg/psycopg2#490]
When trying to use _psycopg2-static_[1] I get the following error. Not static enough?

    Unable to import module 'testscript': /var/task/psycopg2/_psycopg.so: undefined symbol: GSS_C_NT_HOSTBASED_SERVICE

The _lambda-packages_[2] repository works just fine, except that it's weirdly packaged in that it has .tar.gz files of all the modules, so it doesn't fit well into automatic packaging pipelines. Added to that it feels safer to depend on an official psycopg2 module.

1. https://pypi.python.org/pypi/psycopg2-static/
2. https://github.com/Miserlou/lambda-packages

**Comment by dvarrazzo** [comment]
I have never heard of the `psycopg2-static package` before. It's not an official thing. It could be malware as far as I know. I think the presence of that package should be reported to the PyPI maintainer.

**Comment by dsvensson** [comment]
My worries as well. I'll see what I can do. If it's possible to get it nuked from orbit, could you start providing a statically linked replacement?

**Comment by dvarrazzo** [comment]
Just managed to find the email of the author, I'll contact him. Of course I don't assume malice: just an uncoordinated attempt to solve a problem.

**Comment by Jc2k** [comment]
Hi chaps; just got your message. 

@dsvensson - I had the same issues with lambda-packages as yourself, which is why i looked into a wheel version of pyscopg2. I've nuked it from orbit. Certainly no malice, I had hoped to get something that worked and then feed it back upstream. I'm aware of the challenges posed by manylinux wheels (#425) so wanted to avoid causing any fuss for anyone until i had something that worked.

Unfortunately the day job no longer involves postgres so my time to look at anything like this has passed so it is better of rm'd.

**Comment by dvarrazzo** [comment]
@Jc2k thank you very much, I appreciate your attempt into looking about providing a static/manylinux package.

Closing this bug in this form, but solving the dependencies problem and providing static self-contained binaries is something that probably many would appreciate.

**Comment by dsvensson** [comment]
So new issue for releasing a static version at pypi in parallel to the current package then? Or really.. why have a non-static version at all? Is there any benefit of not making sure the intended libpq version is used? Distros will ofc care about dynamically linking, but I doubt anyone would be sad if pypi had it statically linked?

**Comment by Jc2k** [comment]
The primary issue is indeed that you become responsible for e.g. the ssl bindings (my use case was AWS lambda talking to AWS RDS over SSL) which have a lot of churn and probably have to do extra releases even when pyscopg hasn't changed because of this.

**Comment by dvarrazzo** [comment]
Well, in #425 I've already expressed my doubts (in a ranty way, as usual... I apologise for my tone) about the suitability of a manylinux wheel for psycopg. Dynamic libs are useful precisely so that psycopg, libpq, and openssl libraries can change with their own release needs; in a static build an SSL bug becomes a psycopg security bug.

So it would be nice to have the static wheel in the same PyPI location of the src distribution (how do you tell `pip` which one do you want? I guess there is a way). But that would force making a new bugfix release when the libs change. Having a separate package is an hassle too: a Python program doesn't depend on `psycopg2-static`, it depends on `psycopg2` and the fact that what is installed is static is an implementation detail of the artifact received by PyPI, which should be transparent to the program using it.

Probably the depending libraries don't change so often, so the inconvenience of releasing some spurious bugfix release is a lesser evil than having an entirely separate package. I should parse [pep 440](https://www.python.org/dev/peps/pep-0440/) again, but if 4-digits version number are allowed I'd be fine at releasing a static package with 2.6.2, and then release 2.6.2.1, 2.6.2.2 etc. if there is a need to update a library we depend on.

**Comment by dvarrazzo** [comment]
The PyPI version scheme is:

> A version identifier that consists solely of a release segment and optionally an epoch identifier is termed a "final release".
>
> The release segment consists of one or more non-negative integer values, separated by dots:
>
>        N(.N)*

So, about formal correctness, it would be fine to have a version with a 4th digit, stating the artifact release number.

**Comment by psyhomb** [comment]
https://gitlab.com/psyhomb/aws-psycopg2
